Add proto file

iisuslik43 opened the pull request.

iisuslik43 committed:
Add proto file

dbarashev commented on src/main/proto/cosmas.proto:
```diff
@@ -0,0 +1,16 @@
+message Text {
```
Some legal stuff: please add a copyright and author lines at the very top:
/**
Copyright 2018 BarD Software s.r.o

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

dbarashev commented on src/main/proto/cosmas.proto:
```diff
@@ -0,0 +1,16 @@
+message Text {
```
We use a naming convention where request protocol buffers are suffixed with `Request` and response protocol buffers are suffixed with `Response`, and their base name is normally the name of the method. If your method is `createNewVersion` then response protocol buffer is `CreateNewVersionResponse`. Please rename this and other messages appropriately.

dbarashev replied:
Please also add `syntax = "proto3";` line at the very top, because you are actually using proto3 synatx.

dbarashev commented on src/main/proto/cosmas.proto:
```diff
@@ -0,0 +1,16 @@
+message Text {
+  string text = 0
```
[From the docs](https://developers.google.com/protocol-buffers/docs/proto3#assigning-tags):
> The smallest tag number you can specify is 1

dbarashev commented:
Hey, Alexander. 
The purpose of issue #1 was not building something real, the purpose was to start creating build infrastructure (gradle files, protocol buffer plugins, all that stuff), that's why I asked about a simple HelloWorld service. This one is also okay as something to start with, though. The real service will be more complex than this one, so I will not even bother with writing comments about message semantics.

But.
 
I thought you would be curious to compile code, but if that was not obvious, let me say it explicitly: pull request does not have to be completed, but it **has to be compilable**. You reviewer may or may not compile it, but if it is not compilable, it is your fault. Meanwhile:

```
# protoc --java_out=. cosmas.proto
cosmas.proto:5:1: Expected ";".
cosmas.proto:9:3: Expected ";".
cosmas.proto:15:35: Expected type name.
cosmas.proto:16:25: Expected type name.
cosmas.proto:17:19: Expected message type.
```

iisuslik43 committed:
Add gradle build, that compiles all files .proto a
nd puts result into build/generated/source/proto/main/java.
This genereted java class can be used later.

dbarashev commented on build.gradle:
```diff
@@ -0,0 +1,65 @@
+    }
+    plugins {
+        grpc {
+            artifact = 'io.grpc:protoc-gen-grpc-java:1.0.0-pre2'
```
1.0 pre2 version is very outdated. [1.10.0 is the most recent](https://mvnrepository.com/artifact/io.grpc/protoc-gen-grpc-java). You can try using 1.+ syntax to make dependency management system pulling the most recent version from 1.x.x family.

The same applies to grpc dependencies above

dbarashev commented on build.gradle:
```diff
@@ -0,0 +1,65 @@
+group 'iisuslik43.ru'
```
:) Please change this to `com.bardsoftware.papeeria.backend.cosmas`

dbarashev commented on src/main/proto/cosmas.proto:
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+
```
I think you need to specify at least `java_package` and `java_outer_classname` options ([see the docs](https://developers.google.com/protocol-buffers/docs/javatutorial))

Java package is `com.bardsoftware.papeeria.backend.cosmas`, outer class name `CosmasProto`

iisuslik43 replied:
done

dbarashev commented on src/test/kotlin/ru/iisuslik/ProtoTest.kt:
```diff
@@ -0,0 +1,13 @@
+package ru.iisuslik
```
Move this into `com.bardsoftware.papeeria.backend.cosmas` package as well

dbarashev commented on build.gradle:
```diff
@@ -0,0 +1,65 @@
+    }
+    dependencies {
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+        classpath 'com.google.protobuf:protobuf-gradle-plugin:0.8.4'
```
Replace version with `0.8.+` to upgrade to the latest versions automatically

dbarashev commented on src/test/kotlin/ru/iisuslik/ProtoTest.kt:
```diff
@@ -0,0 +1,13 @@
+class ProtoTest {
+    @Test
+    fun cosmasOuterClassHasBeenCompiled() {
+        assertEquals("CosmasOuterClass${'$'}VersionResponse",
```
This test is nice, but apparently it tests Java naming conventions for nested classes :) The fact that protocol buffer has compiled and is available in the classpath is tested merely by accessing it in the second argument of `assertEquals`. 

If you just want a simple test, write `assertNotNull(CosmasOuterClass.VersionResponse.getDefaultInstance())`

dbarashev reviewed (commented):
Looks way better! I am getting this when running `gradle build ` though:
```
e: /tmp/papeeria-edit-history/src/test/kotlin/ru/iisuslik/ProtoTest.kt: (11, 34): Unresolved reference: VersionResponse

```

dbarashev commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.0-bin.zip
```
You may want to specify the latest Gradle 4.6 version.

iisuslik43 committed:
All libraries versions are the newest now, files live in correct packages.

dbarashev commented on src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/ProtoTest.kt:
```diff
@@ -0,0 +1,13 @@
+package com.bardsoftware.papeeria.backend.cosmas
```
Please add the same copyright notice as in `cosmas.proto` in the very beginning of this file

iisuslik43 replied:
done

dbarashev commented on src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/ProtoTest.kt:
```diff
@@ -0,0 +1,13 @@
+
+import org.junit.Assert.*
+
+class ProtoTest {
```
Add a short doc-comment and `@author` tag with your real name and github account:
```
/**
* This class <insert here a brief description of the purpose of this class>
* @author Alexander Fedotov (iisuslik43)
*/
```

iisuslik43 replied:
done

dbarashev reviewed (commented):
Cool, we are almost done.

iisuslik43 committed:
Add docs and copyright notice

iisuslik43 committed:
Change group in build.gradle

dbarashev reviewed (approved):
[LGTM](https://www.urbandictionary.com/define.php?term=LGTM)
Congrats with passing the first code review! You can now merge this pull request.
